### ISSUE TITLE
Design: rethink how we handle interoperability tests in start_test

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#12507]
This was motivated by the tradeoffs of LLVM --library testing, but likely applies to general --library testing.

Today, testing --library involves both a C/Python file and a Chapel file.  Often both of the files involve require their own compilation flags and sometimes their own execution flags.  We can run C files through `start_test` (so `--library` tests with C typically notest the .chpl file and compile it using a precomp on the C file) but can't do the same for Python (so `--library` tests with Python are run using a prediff on the Chapel file).  The C set up means that `start_test -compopts` commands don't get applied when compiling the Chapel files, because we don't propagate them to the .precomp script.  But rearranging testing so that the C files get compiled when testing the Chapel file would potentially tie us to using a particular C compiler (instead of allowing us to run the tests with gcc/icc/cce/etc).

I think probably the best solution is to make the testing system aware of the connection between these tests in some way, so that the Chapel files can adjust to `start_test`'s provided arguments and the C files can get compiled with the relevant `CHPL_TARGET_COMPILER`.  Hopefully this would also reduce the number of files needed to test this interaction.

Note: the YAML based testing configuration solution might be helpful in this regard or at least should keep this issue in mind.

**Comment by lydia-duncan** [comment]
Note that while this support is not present, we have to maintain multiple directories to enable testing `--library` with `--llvm`.  This seems like a maintenance burden going forward

**Comment by mppf** [comment]
> The C set up means that start_test -compopts commands don't get applied when compiling the Chapel files, because we don't propagate them to the .precomp script.

What about simply checking for a COMPOPTS environment variable and using it? Is there a reason that doesn't work?

**Comment by lydia-duncan** [comment]
Based on #11009, I suspect that won't work, but I didn't actually check myself (we might divide things into C/Chapel COMPOPTS during start_test's run in a way that is separate from the COMPOPTS files)

**Comment by mppf** [comment]
Even so, couldn't start_test just make e.g. `CHPL_COMPOPTS` be an environment variable of the current Chapel compopts settings?

**Comment by lydia-duncan** [comment]
We could do that!  Were you thinking this would involve solving #11009 or just doing part of it?  I'd rather we didn't do a partial solution if we could avoid it

**Comment by mppf** [comment]
I don't think we need to solve #11009 to do this - the tests in questions could just not use directory-level compopts. (I think it's reasonable to solve #11009 - I just don't think it's required for this issue).
